helix-tui/src/buffer.rs

```diff
@@ -583,6 +583,50 @@ impl Buffer {
         }
         updates
     }
+
+    /// Apply shade to all cells in area.
+    ///
+    /// shade = 0 -> set text modifier DIM
+    /// shade < 0 -> darken rgb color
+    /// shade > 0 -> lighten rgb color
+    pub fn dim(&mut self, area: Rect, shade: i8) {
+        let shade_amount = shade.unsigned_abs().saturating_mul(2);
+        let shade_fn = if shade < 0 {
+            u8::saturating_sub
+        } else {
+            u8::saturating_add
+        };
+        // TODO: replace with saturating_add_signed when mixed_integer_ops (https://github.com/rust-lang/rust/issues/87840) get stabilized.
+        let shaded = |v| shade_fn(v, shade_amount);
```

Comment:
I'd like to note that saturating addition/subtraction has a very different effect than that of [`palette::Shade::{lighten, darken}`](https://docs.rs/palette/0.6.0/src/palette/rgb/rgb.rs.html#535-580), which was used before 9ad01a06232ae03ed99cbd601764aab29fe5b379, and is more likely to cause undesirable effects such as clipping, saturation boosting, and hue shifting (examples shown below).

- ![](https://img.shields.io/badge/-%E3%80%80-64460a?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-0a6400?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-325014?style=for-the-badge) `[(100, 70, 10), (10, 100, 0), (50, 80, 20)]` (various dark tones) → ![](https://img.shields.io/badge/-%E3%80%80-000?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-000?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-000?style=for-the-badge) `[(0, 0, 0); 3]` (black)
- ![](https://img.shields.io/badge/-%E3%80%80-c86464?style=for-the-badge) `(200, 100, 100)` (desaturated red) → ![](https://img.shields.io/badge/-%E3%80%80-640000?style=for-the-badge) `(100, 0, 0)` (darker, but more vivid red)
- ![](https://img.shields.io/badge/-%E3%80%80-c86400?style=for-the-badge) `(200, 100, 0)` (orange) → ![](https://img.shields.io/badge/-%E3%80%80-640000?style=for-the-badge) `(100, 0, 0)` (dark red)

Here's a possible implementation based on alpha compositing without gamma correction (warning: untested):

```suggestion
        let alpha = i32::from(shade).unsigned_abs();
        let (src_factor, dst_factor) = (alpha, 256 - alpha);
        let src = if shade > 0 { 255u32 } else { 0u32 } * src_factor;
        let shaded = |dst_color: u8| ((u32::from(dst_color) * dst_factor + src) >> 8) as u8;
```

... which should produce colors like these:

- ![](https://img.shields.io/badge/-%E3%80%80-64460a?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-0a6400?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-325014?style=for-the-badge) `[(100, 70, 10), (10, 100, 0), (50, 80, 20)]` (various dark tones) → ![](https://img.shields.io/badge/-%E3%80%80-322305?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-053200?style=for-the-badge)![](https://img.shields.io/badge/-%E3%80%80-19280a?style=for-the-badge) `[(50, 35, 5), (5, 50, 0), (25, 40, 10)]` (various darker tones)
- ![](https://img.shields.io/badge/-%E3%80%80-c86464?style=for-the-badge) `(200, 100, 100)` (desaturated red) → ![](https://img.shields.io/badge/-%E3%80%80-643232?style=for-the-badge) `(100, 50, 50)` (darker, still desaturated red)
- ![](https://img.shields.io/badge/-%E3%80%80-c86400?style=for-the-badge) `(200, 100, 0)` (orange) → ![](https://img.shields.io/badge/-%E3%80%80-643200?style=for-the-badge) `(100, 50, 0)` (dark orange)

Reply:
This works great. Thank you 🙏